### PR TITLE
Pr15 split part3

### PR DESCRIPTION
**Pull request by qinzhang22**
This is removing the use 'file' and updating 'xmlrpc, socketserver, urllib' syntax for python3. For this part, we didn't consider python2 compatibility considering it will make code complicated.

### CODE REVIEW

**Comment by andyhhp** [comment]
The first patch (`file`) is fine, and could have been part of the previous misc if I'd noticed it.

However, the second patch is definitely not Py2-safe, so we can't take it in that form.  See:

https://python-future.org/compatible_idioms.html#urllib-module
https://python-future.org/compatible_idioms.html#xmlrpc-module

File: XSConsoleRemoteTest.py
Comment (bernhardkaindl) [inline XSConsoleRemoteTest.py:16]:
```suggestion
import os
import socket
```

File: XSConsoleRemoteTest.py
Comment (bernhardkaindl) [reply XSConsoleRemoteTest.py:25]: I advocate for this simple solution to replace this entire commit with these two hunks.

The main reason is that the the python-future module became unmaintained and projects are already forced to drop their use of python-future because of this when they want to support Python 3.12. One Example:

https://github.com/gkovacs/lz-string-python/pull/6/files

A fixup for this particular issue of future using `import imp` which has been dropped has been submitted 3 weeks ago, but the maintainers of future have gone missing in January, which is now 9 months without any commit or merge. The deprecation warning from the imp module has been spread wide from the stable release of 3.7 to 3.11, with already 3.10 specifically warning that  imp will not be killed with 3.12, which it did:

```py
# python3.10 -Wdefault -c 'from future import standard_library;standard_library.install_aliases()'
/usr/lib/python3/dist-packages/future/standard_library/__init__.py:65: DeprecationWarning: the imp module is deprecated in favour of importlib and slated for removal in Python 3.12; see the module's documentation for alternative uses
  import imp
# python3.12 -c 'from future import standard_library;standard_library.install_aliases()'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3/dist-packages/future/standard_library/__init__.py", line 65, in <module>
    import imp
ModuleNotFoundError: No module named 'imp'
```
Therefore, just use the simple try: import method and no change to the actual code.

That's fully in line with Andrew's demand for Python2 compat and it avoids having dependency on no longer maintained packages (there is also six.moves, but that also is a not needed dependency which can become unmaintained). Besides, when pylint would used (may be good when developing new code), two pylint warnings would have to be disabled, because installing the aliases would have to be put on top, violating the import order:
```py
# pylint: disable=wrong-import-position,wrong-import-order
from future import standard_library
standard_library.install_aliases()
```

**Comment by andyhhp** [comment]
> and with consideration that future is not maintained anymore and is broken in Python3.12

Lovely...  I'll defer to your suggestion for how best to do this.

**Comment by liulinC** [comment]
Again, I want to highlight, 
XS8 already has python3 (3.6), so **pure python3 code is compatible for XS9 and XS8 stream**. (That is also the goal of the Epic to remove python2 from dom0),   **Note**: XSConsole is not a lib like xen, or scp-python-libs, It does not need to provides py2 version for other callers.

The only usage of py2 of XSConsole is for Yangtze, which will be EoL in 2025.
Considering this repo only updates several times during years, I have good reason we does not need much p2 updates.
If we do need to update Yangtze, then branch out for Yangtze support.

On the other side, to make master compatible for py2, we make the code unnecessary complicated.
@andyhhp @rosslagerwall @bernhardkaindl

**Comment by andyhhp** [comment]
Hmm, on second thoughts, you're right.  There's no dependency of xsconsole that's going to need it to stay py2 compatible, and we're not going to be backporting anything but critical fixes to ytz.  (Need to double check the way branches are handled in the spec repo, but I expect that will be fine.)

So, new plan, move for Py3.6 baseline.  However, `master` needs to remain atomically py2 or py3 compatible, and not a mix of fixes which leaves it broken in both.

So, please can we collect together all commits which are py2-safe fixes, which can get reviewed and merged in due course, and please keep all commits which break compatibility with py2 in a separate PR so it can be merged all in one go.  The final commit on the py3 branch should bump the major version number of xsconsole seeing as its a bit step change.

Sound ok?

**Comment by liulinC** [comment]
> Hmm, on second thoughts, you're right. There's no dependency of xsconsole that's going to need it to stay py2 compatible, and we're not going to be backporting anything but critical fixes to ytz. (Need to double check the way branches are handled in the spec repo, but I expect that will be fine.)
> 
> So, new plan, move for Py3.6 baseline. However, `master` needs to remain atomically py2 or py3 compatible, and not a mix of fixes which leaves it broken in both.
> 
> So, please can we collect together all commits which are py2-safe fixes, which can get reviewed and merged freely, and please keep all commits which break compatibility with py2 in a separate single PR so it can be merged all in one go. The final commit on the py3 branch should bump the major version number of xsconsole seeing as its a bit step change.
> 
> Sound ok?

Sounds a good and reasonable plan for me @qinzhang22  your thoughts?

**Comment by qinzhang22** [comment]
> > Hmm, on second thoughts, you're right. There's no dependency of xsconsole that's going to need it to stay py2 compatible, and we're not going to be backporting anything but critical fixes to ytz. (Need to double check the way branches are handled in the spec repo, but I expect that will be fine.)
> > So, new plan, move for Py3.6 baseline. However, `master` needs to remain atomically py2 or py3 compatible, and not a mix of fixes which leaves it broken in both.
> > So, please can we collect together all commits which are py2-safe fixes, which can get reviewed and merged freely, and please keep all commits which break compatibility with py2 in a separate single PR so it can be merged all in one go. The final commit on the py3 branch should bump the major version number of xsconsole seeing as its a bit step change.
> > Sound ok?
> 
> Sounds a good and reasonable plan for me @qinzhang22 your thoughts?

**Comment by qinzhang22** [comment]
I updated the PR per our agreement that collecting together all commits which are py2-safe fixes. Hopefully it can be reviewed and merged freely.

File: plugins-base/XSFeatureSRInfo.py
Comment (andyhhp) [inline plugins-base/XSFeatureSRInfo.py:115]: is `-sr` right here?

Personally, I think the logic would be easier to follow as

```python
# Sort list by SR shared flag then name
srList.sort(key=lambda sr: sr.name_label(''))
srList.sort(key=lambda sr: sr.shared(False))
```
maybe with a `reverse=True` in the second line if I've interpreted the intent of the negation properly ?

**Review by andyhhp — COMMENTED** [review]
Everything else LGTM.

File: plugins-base/XSFeatureSRInfo.py
Comment (andyhhp) [reply plugins-base/XSFeatureSRInfo.py:115]: Alternatively, `srList.sort(key=lambda sr: (not sr.shared(False), sr.name_label('')))` might be better ?  At least it makes clear that the first part of the tuple is supposed to be a boolean.

Tangent, passing the default-if-missing value in makes this code horrible to read and follow.  Do you happen to know if this is a local artifact, or a property of the API?

Comment (bernhardkaindl) [reply plugins-base/XSFeatureSRInfo.py:115]: > is -sr right here?
Yes, it is correct. The 1st half of the original `cmp(y.shared(False), x.shared(False))` had y and x reversed, so the primary sort order is reversed.

Splitting the nested sort into two consecutive sorts would make the result reliant of the sort algorithm chosen by the Python interpreter for the given data and I'd like the result to be kept clearly defined as a nested sort with two sort keys. Therefore, I see the change is correct!

Because the sort is already commented using a meaningful comment, the sort does not need further explanation:
```py
# Sort list by SR shared flag then name
srList.sort(key=lambda sr: (-sr.shared(False), sr.name_label('')))
```
Even though the comment says that `sr.shared(False)` returns a flag, nothing here ensures that it will always return a boolean. In principle, it could return anything (numbers, floats or a mix of all that). Therefore I think keeping `-sr.shared(False)` is the safe thing to do (does not change the behavior at all) and it's clear that `-` means negation.

**Review by bernhardkaindl — APPROVED** [review]
Reviewed the changes and checked the sort lambdas.

**Review by liulinC — APPROVED** [review]
Approve basing on my review, and the discussion.

**Pull request merged**
